pinecone/client.go

```diff
@@ -20,3 +41,357 @@ func (c *Client) IndexWithNamespace(host string, namespace string) (*IndexConnec
+			Shards           *control.PodSpecShards   `json:"shards,omitempty"`
+			SourceCollection *string                  `json:"source_collection,omitempty"`
+		}{
+			Environment:      in.Environment,
+			PodType:          in.PodType,
+			Pods:             in.TotalCount(),
+			Replicas:         in.ReplicaCount(),
+			Shards:           in.ShardCount(),
+			SourceCollection: in.SourceCollection,
+		},
+	}
+	if in.MetadataConfig != nil {
+		req.Spec.Pod.MetadataConfig = &struct {
+			Indexed *[]string `json:"indexed,omitempty"`
+		}{
+			Indexed: in.MetadataConfig.Indexed,
+		}
+	}
+
+	res, err := c.restClient.CreateIndex(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusCreated {
+		var errResp control.ErrorResponse
+		err = json.NewDecoder(res.Body).Decode(&errResp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode error response: %w", err)
+		}
+		return nil, fmt.Errorf("failed to create index: %s", errResp.Error.Message)
+	}
+
+	return decodeIndex(res.Body)
+}
+
+type CreateServerlessIndexRequest struct {
+	Name      string
+	Dimension int32
+	Metric    IndexMetric
+	Cloud     Cloud
+	Region    string
+}
+
+func (c *Client) CreateServerlessIndex(ctx context.Context, in *CreateServerlessIndexRequest) (*Index, error) {
+	metric := control.IndexMetric(in.Metric)
+	req := control.CreateIndexRequest{
+		Name:      in.Name,
+		Dimension: in.Dimension,
+		Metric:    &metric,
+		Spec: control.CreateIndexRequest_Spec{
+			Serverless: &control.ServerlessSpec{
+				Cloud:  control.ServerlessSpecCloud(in.Cloud),
+				Region: in.Region,
+			},
+		},
+	}
+
+	res, err := c.restClient.CreateIndex(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusCreated {
+		var errResp control.ErrorResponse
+		err = json.NewDecoder(res.Body).Decode(&errResp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode error response: %w", err)
+		}
+		return nil, fmt.Errorf("failed to create index: %s", errResp.Error.Message)
+	}
+
+	return decodeIndex(res.Body)
+}
+
+func (c *Client) DescribeIndex(ctx context.Context, idxName string) (*Index, error) {
+	res, err := c.restClient.DescribeIndex(ctx, idxName)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		var errResp control.ErrorResponse
+		err = json.NewDecoder(res.Body).Decode(&errResp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode error response: %w", err)
+		}
+		return nil, fmt.Errorf("failed to describe idx: %s", errResp.Error.Message)
+	}
+
+	return decodeIndex(res.Body)
+}
+
+func (c *Client) DeleteIndex(ctx context.Context, idxName string) error {
+	res, err := c.restClient.DeleteIndex(ctx, idxName)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusAccepted {
+		var errResp control.ErrorResponse
+		err = json.NewDecoder(res.Body).Decode(&errResp)
```

Comment:
Is there a potential resource leak here of the response body not being consumed? 

From https://pkg.go.dev/net/http#:~:text=If%20the%20returned,keep%2Dalive%22%20request.

> If the returned error is nil, the [Response](https://pkg.go.dev/net/http#Response) will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the [Client](https://pkg.go.dev/net/http#Client)'s underlying [RoundTripper](https://pkg.go.dev/net/http#RoundTripper) (typically [Transport](https://pkg.go.dev/net/http#Transport)) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

Reply:
The defer line 217 takes care of closing the body. The decoder actually consumes the entire body as well.

Reply:
The status code isn't 202. The body will be a JSON document in the shape of `control.ErrorResponse`. Line 221 passes the `res.Body` to a json decoder and a reference to `errResp` to hold the parsed body. When the body is parsed, it is read.

Reply:
Sorry I wasn't clear -- should have selected the larger code block including return. I am actually asking about the case where it falls through because the Status is not 202 and just returns nil

Reply:
I think the `defer res.Body.Close()` deals with it either way, right?